spec/unit/transaction_spec.rb

```diff
@@ -420,6 +420,19 @@ def generate
       transaction.evaluate
     end
 
+    it "should call ::pre_resource_eval on provider classes that support it" do
+      skip if Puppet::Util::Platform.windows?
+      selinux = double('selinux', is_selinux_enabled: true)
+      stub_const('Selinux', selinux)
+
+      resource = Puppet::Type.type(:file).new(:path => make_absolute("/tmp/foo"))
+      transaction = transaction_with_resource(resource)
+
+      expect(resource.provider.class).to receive(:pre_resource_eval)
```

Comment:
I think this will fail since the file provider doesn't implement `pre_resource_eval` (yet). I would instead create a new type and provider in `transaction_spec.rb` like we do when testing `post_resource_eval` (but change the `teardown_test` and `teardown_provider` names to something else)

https://github.com/puppetlabs/puppet/blob/f5a464e0f3319533dc144d4d965fb4a7bdb2543b/spec/unit/transaction_spec.rb#L700-L716

Then write a test to verify the `pre_resource_eval` method get called, similar to https://github.com/puppetlabs/puppet/blob/f5a464e0f3319533dc144d4d965fb4a7bdb2543b/spec/unit/transaction_spec.rb#L718-L725